### PR TITLE
Add downgrade prevention for swaps

### PR DESCRIPTION
**Pull request by kasjer**
Currently, downgrade prevention was limited to overwrite only
builds (version check) or devices with hardware storage for
security counter.

This extends downgrade prevention to be used when swap update
is selected.
Unlike MCUBOOT_HW_ROLLBACK_PROT option it does not require user
code to provide external way to store security counter.
Security counter from slot 1 image is used for comparison.
With security counter usage it is possible to have limited
software rollback if security counter was not incremented.

It is possible to use image version where strict rule for
image version comparison prevents any downgrades.

Downgrade prevention is also added to mynewt configuration.

If image in slot 1 is marked as pending and downgrade prevention
is in place, image will be deleted to avoid check on next boot.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

### CODE REVIEW

File: boot/bootutil/src/loader.c [hunk 1]
```diff
@@ -1902,6 +1902,43 @@ boot_update_hw_rollback_protection(struct boot_loader_state *state)
...
+static int
+check_downgrade_prevention(struct boot_loader_state *state)
+{
+#if defined(MCUBOOT_DOWNGRADE_PREVENTION) && (defined(MCUBOOT_SWAP_USING_MOVE) || defined(MCUBOOT_SWAP_USING_SCRATCH))
+    uint32_t security_counter[2];
+    int rc;
+
+    if (MCUBOOT_DOWNGRADE_PREVENTION_SECURITY_COUNTER) {
+        /* If there was security no counter in slot 0, allow swap */
+        rc = bootutil_get_img_security_cnt(&(BOOT_IMG(state, 0).hdr), BOOT_IMG(state, 0).area, &security_counter[0]);
```
Comment (d3zd3z) [inline boot/bootutil/src/loader.c:1914]: Please try to keep lines to 80-90 characters. These lines, including the #if`` above should be split across two or more lines.

File: boot/mynewt/mcuboot_config/include/mcuboot_config/mcuboot_config.h [hunk 2]
```diff
@@ -85,6 +85,16 @@
 #if MYNEWT_VAL(BOOTUTIL_BOOTSTRAP)
 #define MCUBOOT_BOOTSTRAP 1
 #endif
+#if MYNEWT_VAL_CHOICE(BOOTUTIL_DOWNGRADE_PREVENTION, version)
+#define MCUBOOT_DOWNGRADE_PREVENTION                     1
```
Comment (d3zd3z) [inline boot/mynewt/mcuboot_config/include/mcuboot_config/mcuboot_config.h:89]: Maybe a comment here. It is not clear that `MCUBOOT_DOWNGRADE_PREVENTION` is used in a `#if defined` construct, whereas `MCUBOOT_DOWNGRADE_PREVENTION_SECURITY_COUNTER` is used as a boolean.

**Review by d3zd3z — COMMENTED** [review]
It would be nice to also have support in the Zephyr kconfig files.

File: boot/bootutil/src/loader.c (hunk 1 above)
Comment (kasjer) [reply boot/bootutil/src/loader.c]: lines length limited as requested

File: boot/mynewt/mcuboot_config/include/mcuboot_config/mcuboot_config.h (hunk 2 above)
Comment (kasjer) [reply boot/mynewt/mcuboot_config/include/mcuboot_config/mcuboot_config.h:89]: comment added

**Comment by kasjer** [comment]
> It would be nice to also have support in the Zephyr kconfig files.

zephyr Kconfig updated as requested

**Review by d3zd3z — APPROVED** [review]

**Comment by sjanc** [comment]
gentle ping :)

**Pull request merged**